### ISSUE TITLE
Errors encountered in model training

### ISSUE THREAD

**Issue by A12-RUI** [tensorflow/tensorflow#39501]
I encountered an errors when I was training mask_rcnn model with my own dataset .  So I ask for you help . I will appreciate it if you can help me out.

Caused by op 'GatherV2_4', defined at:
  File "model_main.py", line 110, in <module>
    tf.app.run()
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\platform\app.py", line 125, in run
    _sys.exit(main(argv))
  File "model_main.py", line 106, in main
    tf.estimator.train_and_evaluate(estimator, train_spec, eval_specs[0])
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\estimator\training.py", line 471, in train_and_evaluate
    return executor.run()
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\estimator\training.py", line 610, in run
    return self.run_local()
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\estimator\training.py", line 711, in run_local
    saving_listeners=saving_listeners)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\estimator\estimator.py", line 354, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\estimator\estimator.py", line 1207, in _train_model
    return self._train_model_default(input_fn, hooks, saving_listeners)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\estimator\estimator.py", line 1237, in _train_model_default
    features, labels, model_fn_lib.ModeKeys.TRAIN, self.config)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\estimator\estimator.py", line 1195, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\model_lib.py", line 308, in model_fn
    features[fields.InputDataFields.true_image_shape])
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\meta_architectures\faster_rcnn_meta_arch.py", line 846, in predict
    true_image_shapes))
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\meta_architectures\faster_rcnn_meta_arch.py", line 1016, in _predict_second_stage
    image_shape, true_image_shapes)
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\meta_architectures\faster_rcnn_meta_arch.py", line 746, in _proposal_postprocess
    anchors, image_shape_2d, true_image_shapes)
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\meta_architectures\faster_rcnn_meta_arch.py", line 1700, in _postprocess_rpn
    groundtruth_weights_list)
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\meta_architectures\faster_rcnn_meta_arch.py", line 1786, in _sample_box_classifier_batch
    single_image_groundtruth_weights)
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\meta_architectures\faster_rcnn_meta_arch.py", line 1914, in _sample_box_classifier_minibatch_single_image
    groundtruth_weights=groundtruth_weights)
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\core\target_assigner.py", line 196, in assign
    reg_weights = self._create_regression_weights(match, groundtruth_weights)
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\core\target_assigner.py", line 334, in _create_regression_weights
    groundtruth_weights, ignored_value=0., unmatched_value=0.)
  File "C:\Users\瑞\Documents\GitHub\models\research\object_detection\core\matcher.py", line 214, in gather_based_on_match
    gathered_tensor = self._gather_op(input_tensor, gather_indices)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\ops\array_ops.py", line 2675, in gather
    return gen_array_ops.gather_v2(params, indices, axis, name=name)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\ops\gen_array_ops.py", line 3331, in gather_v2
    "GatherV2", params=params, indices=indices, axis=axis, name=name)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\util\deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\framework\ops.py", line 3274, in create_op
    op_def=op_def)
  File "C:\ProgramData\Anaconda3\envs\first_keras\lib\site-packages\tensorflow\python\framework\ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): indices[0] = 0 is not in [0, 0)
         [[node GatherV2_4 (defined at C:\Users\瑞\Documents\GitHub\models\research\object_detection\core\matcher.py:214)  = GatherV2[Taxis=DT_INT32, Tindices=DT_INT64, Tparams=DT_FLOAT, _device="/device:CPU:0"](cond/Merge, Reshape_8, GatherV2_3/axis)]]
         [[node IteratorGetNext (defined at model_main.py:106)  = IteratorGetNext[output_shapes=[[1], [1,?,?,3], [1,2], [1,3], [1,100], [1,100,4], [1,100,4], [1,100,4], [1,100], [1,100,?,?], [1,100], [1,100], [1]], output_types=[DT_INT32, DT_FLOAT, DT_INT32, DT_INT32, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_INT32, DT_FLOAT, DT_BOOL, DT_FLOAT, DT_INT32], _device="/job:localhost/replica:0/task:0/device:CPU:0"](IteratorV2)]]

**Comment by amahendrakar** [comment]
@A12-RUI,
In order to expedite the trouble-shooting process, could you please provide the complete code, the dataset and the TensorFlow version you are using. Thanks!

**Comment by A12-RUI** [comment]
I used  coco type dataste ,and tensorflow version is 1.12.0.
code:
# coding: utf-8

# # Object Detection Demo
# Welcome to the object detection inference walkthrough!  This notebook will walk you step by step through the process of using a pre-trained model to detect objects in an image. Make sure to follow the [installation instructions](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/installation.md) before you start.


from distutils.version import StrictVersion
import numpy as np
import os
import six.moves.urllib as urllib
import sys
import tarfile
import tensorflow as tf
import zipfile
from IPython.display import display
from collections import defaultdict
from io import StringIO
from matplotlib import pyplot as plt
from PIL import Image

# This is needed since the notebook is stored in the object_detection folder.

sys.path.append(r'C:\Users\瑞\Documents\GitHub\models\research')
from object_detection.utils import ops as utils_ops

# if StrictVersion(tf.__version__) < StrictVersion('1.9.0'):
#   raise ImportError('Please upgrade your TensorFlow installation to v1.9.* or later!')




# This is needed to display the images.
# get_ipython().magic(u'matplotlib inline')


# ## Object detection imports
# Here are the imports from the object detection module.



from utils import label_map_util

from utils import visualization_utils as vis_util


# # Model preparation

# ## Variables
#
# Any model exported using the `export_inference_graph.py` tool can be loaded here simply by changing `PATH_TO_FROZEN_GRAPH` to point to a new .pb file.
#
# By default we use an "SSD with Mobilenet" model here. See the [detection model zoo](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md) for a list of other models that can be run out-of-the-box with varying speeds and accuracies.

# What model to download.
MODEL_NAME = 'C:/Users/瑞/Documents/GitHub/models/research/object_detection/test_defects_detection'   ###############
# MODEL_FILE = MODEL_NAME + '.tar.gz'
# DOWNLOAD_BASE = 'http://download.tensorflow.org/models/object_detection/'

# Path to frozen detection graph. This is the actual model that is used for the object detection.
PATH_TO_FROZEN_GRAPH = MODEL_NAME + '/frozen_inference_graph.pb'

# List of the strings that is used to add correct label for each box.
PATH_TO_LABELS = os.path.join('data', 'test_label_map.pbtxt')  ################

NUM_CLASSES = 4  ######################


# ## Download Model



# opener = urllib.request.URLopener()
# opener.retrieve(DOWNLOAD_BASE + MODEL_FILE, MODEL_FILE)
'''
tar_file = tarfile.open(MODEL_FILE)
for file in tar_file.getmembers():
  file_name = os.path.basename(file.name)
  if 'frozen_inference_graph.pb' in file_name:
    tar_file.extract(file, os.getcwd())
'''

# ## Load a (frozen) Tensorflow model into memory.



detection_graph = tf.Graph()
with detection_graph.as_default():
  od_graph_def = tf.GraphDef()
  with tf.gfile.GFile(PATH_TO_FROZEN_GRAPH, 'rb') as fid:
    serialized_graph = fid.read()
    od_graph_def.ParseFromString(serialized_graph)
    tf.import_graph_def(od_graph_def, name='')


# ## Loading label map
# Label maps map indices to category names, so that when our convolution network predicts `5`, we know that this corresponds to `airplane`.  Here we use internal utility functions, but anything that returns a dictionary mapping integers to appropriate string labels would be fine



label_map = label_map_util.load_labelmap(PATH_TO_LABELS)
categories = label_map_util.convert_label_map_to_categories(label_map, max_num_classes=NUM_CLASSES, use_display_name=True)
category_index = label_map_util.create_category_index(categories)



def load_image_into_numpy_array(image):
  (im_width, im_height) = image.size
  return np.array(image.getdata()).reshape(
      (im_height, im_width, 3)).astype(np.uint8)


# # Detection



# For the sake of simplicity we will use only 2 images:
# image1.jpg
# image2.jpg
# If you want to test the code with your images, just add path to the images to the TEST_IMAGE_PATHS.
PATH_TO_TEST_IMAGES_DIR = 'C:/Users/瑞/Documents/GitHub/models/research/object_detection/defects_test_images'    ####################
#TEST_IMAGE_PATHS = [ os.path.join(PATH_TO_TEST_IMAGES_DIR, '{}.jpg'.format(i)) for i in range(1, 11) ]
##########################
os.chdir(PATH_TO_TEST_IMAGES_DIR)
TEST_IMAGE_PATHS = os.listdir(PATH_TO_TEST_IMAGES_DIR)
# Size, in inches, of the output images.
IMAGE_SIZE = (12, 8)


def run_inference_for_single_image(image, graph):
  with graph.as_default():
    with tf.Session() as sess:
      # Get handles to input and output tensors
      ops = tf.get_default_graph().get_operations()
      all_tensor_names = {output.name for op in ops for output in op.outputs}
      tensor_dict = {}
      for key in [
          'num_detections', 'detection_boxes', 'detection_scores',
          'detection_classes', 'detection_masks'
      ]:
        tensor_name = key + ':0'
        if tensor_name in all_tensor_names:
          tensor_dict[key] = tf.get_default_graph().get_tensor_by_name(
              tensor_name)
      if 'detection_masks' in tensor_dict:
        # The following processing is only for single image
        detection_boxes = tf.squeeze(tensor_dict['detection_boxes'], [0])
        detection_masks = tf.squeeze(tensor_dict['detection_masks'], [0])
        # Reframe is required to translate mask from box coordinates to image coordinates and fit the image size.
        real_num_detection = tf.cast(tensor_dict['num_detections'][0], tf.int32)
        detection_boxes = tf.slice(detection_boxes, [0, 0], [real_num_detection, -1])
        detection_masks = tf.slice(detection_masks, [0, 0, 0], [real_num_detection, -1, -1])
        detection_masks_reframed = utils_ops.reframe_box_masks_to_image_masks(
            detection_masks, detection_boxes, image.shape[0], image.shape[1])
        detection_masks_reframed = tf.cast(
            tf.greater(detection_masks_reframed, 0.5), tf.uint8)
        # Follow the convention by adding back the batch dimension
        tensor_dict['detection_masks'] = tf.expand_dims(
            detection_masks_reframed, 0)
      image_tensor = tf.get_default_graph().get_tensor_by_name('image_tensor:0')

      # Run inference
      output_dict = sess.run(tensor_dict,
                             feed_dict={image_tensor: np.expand_dims(image, 0)})

      # all outputs are float32 numpy arrays, so convert types as appropriate
      output_dict['num_detections'] = int(output_dict['num_detections'][0])
      output_dict['detection_classes'] = output_dict[
          'detection_classes'][0].astype(np.uint8)
      output_dict['detection_boxes'] = output_dict['detection_boxes'][0]
      output_dict['detection_scores'] = output_dict['detection_scores'][0]
      if 'detection_masks' in output_dict:
        output_dict['detection_masks'] = output_dict['detection_masks'][0]
  return output_dict

num=20
for image_path in TEST_IMAGE_PATHS:
  image = Image.open(image_path)
  # the array based representation of the image will be used later in order to prepare the
  # result image with boxes and labels on it.
  image_np = load_image_into_numpy_array(image)
  # Expand dimensions since the model expects images to have shape: [1, None, None, 3]
  image_np_expanded = np.expand_dims(image_np, axis=0)
  # Actual detection.
  output_dict = run_inference_for_single_image(image_np, detection_graph)
  # Visualization of the results of a detection.
  vis_util.visualize_boxes_and_labels_on_image_array(
      image_np,
      output_dict['detection_boxes'],
      output_dict['detection_classes'],
      output_dict['detection_scores'],
      category_index,
      instance_masks=output_dict.get('detection_masks'),
      use_normalized_coordinates=True,
      line_thickness=8)
  plt.imshow(image_np)
  display(Image.fromarray(image_np))
  aaa=Image.fromarray(image_np)
  aaa.save('test00%d.jpg'%num)
  num=num+1

**Comment by amahendrakar** [comment]
@A12-RUI,
Is there any specific reason you are using TF v1.12? 
Could you please upgrade TensorFlow to v1.15 or v2.2 and let us know if you are facing the same issue. Thanks!

**Comment by A12-RUI** [comment]
I have solve the problem ，thanks！
